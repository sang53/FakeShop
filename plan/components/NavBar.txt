inputs (hardcode links & names):
  number of items in shopping cart
  current page? - input from url/router

memo:
  memoise to only re-render if shopping cart changes

output:
  div with links + shopping cart number