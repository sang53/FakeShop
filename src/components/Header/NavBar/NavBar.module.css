.navButton {
  text-decoration: none;
  color: inherit;
  padding: 0.5rem 1rem;
  border-radius: 25px;
  transition: background-color 100ms;

  &:hover {
    background-color: var(--navbar-hover);
  }
}

.active {
  composes: navButton;
  background-color: var(--navbar-active);
}
