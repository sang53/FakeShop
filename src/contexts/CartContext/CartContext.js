import { createContext } from "react";

export const CartItems = createContext(null);
export const CartReducer = createContext(null);
