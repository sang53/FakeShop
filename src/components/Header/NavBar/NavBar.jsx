import { NavLink } from "react-router";
import classes from "./NavBar.module.css";
import { use } from "react";
import { CartItems } from "../../../contexts/CartContext/CartContext";

export default function NavBar() {
  const cartSize = Object.keys(use(CartItems)).length;
  return (
    <nav className={"flex-center no-select " + classes.navBar}>
      <NavLink
        to="/"
        className={({ isActive }) =>
          isActive ? classes.active : classes.navButton
        }
      >
        Home
      </NavLink>
      <NavLink
        to="/categories"
        className={({ isActive }) =>
          isActive ? classes.active : classes.navButton
        }
      >
        Categories
      </NavLink>
      <NavLink
        to="/products"
        className={({ isActive }) =>
          isActive ? classes.active : classes.navButton
        }
      >
        Products
      </NavLink>
      <NavLink
        to="/cart"
        className={({ isActive }) =>
          isActive ? classes.active : classes.navButton
        }
      >
        Cart({cartSize})
      </NavLink>
    </nav>
  );
}
