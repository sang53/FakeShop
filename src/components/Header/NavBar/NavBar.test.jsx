import { test, expect, describe, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";
import NavBar from "./NavBar";
import { CartItems } from "../../../contexts/CartContext/CartContext";
import { MemoryRouter } from "react-router";

const mockCart = {
  1: 3,
  10: 4,
  4: 5,
};

function renderNavBar(value) {
  return render(
    <CartItems.Provider value={value}>
      <MemoryRouter initialEntries={["/"]}>
        <NavBar />
      </MemoryRouter>
    </CartItems.Provider>
  );
}

describe("displays number of unique items in cart", () => {
  test("empty cart", () => {
    renderNavBar({});
    expect(screen.getByRole("link", { name: /cart/i })).toHaveTextContent(
      "Cart(0)"
    );
  });
  test("items in cart", () => {
    renderNavBar(mockCart);
    expect(screen.getByRole("link", { name: /cart/i })).toHaveTextContent(
      "Cart(3)"
    );
  });
});

describe("current page has aria-current", () => {
  beforeEach(() => {
    renderNavBar({});
  });
  test("initial home page", () => {
    const currentPage = screen.getAllByRole("link", { current: "page" });
    expect(currentPage.length).toBe(1);
    expect(currentPage[0].textContent).toMatch(/home/i);
  });
  test("switch to products page", async () => {
    await user.click(screen.getByRole("link", { name: /product/i }));
    expect(screen.getAllByRole("link", { current: "page" }).length).toBe(1);
    expect(screen.getByRole("link", { current: "page" }).textContent).toMatch(
      /product/i
    );
  });
  test("switch to cart page", async () => {
    await user.click(screen.getByRole("link", { name: /cart/i }));
    expect(screen.getAllByRole("link", { current: "page" }).length).toBe(1);
    expect(screen.getByRole("link", { current: "page" }).textContent).toMatch(
      /cart/i
    );
  });
});

test("renders NavBar elements as expected", () => {
  const { container } = renderNavBar({});
  expect(container).toMatchSnapshot();
});
