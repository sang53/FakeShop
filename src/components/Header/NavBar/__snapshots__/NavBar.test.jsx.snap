// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders NavBar elements as expected 1`] = `
<div>
  <nav
    class="flex-center no-select _navBar_1ee360"
  >
    <a
      aria-current="page"
      class="_active_1ee360"
      data-discover="true"
      href="/"
    >
      Home
    </a>
    <a
      class="_navButton_1ee360"
      data-discover="true"
      href="/categories"
    >
      Categories
    </a>
    <a
      class="_navButton_1ee360"
      data-discover="true"
      href="/products"
    >
      Products
    </a>
    <a
      class="_navButton_1ee360"
      data-discover="true"
      href="/cart"
    >
      Cart(
      0
      )
    </a>
  </nav>
</div>
`;
